### ISSUE TITLE
issue compiling on linux

### ISSUE THREAD

**Issue by PHANTOMCLO** [FunkinCrew/Funkin#625]
currently trying to compile on ubuntu 20.04. i followed the readme, but im getting this error.
Warning: Lime version 7.9.0 may not be compatible with OpenFL 9.0.2 (expected version 7.7.*)
Use <lime version="" /> before <haxelib name="openfl" /> to include a specific version
/home/cloavii/haxelib/flixel/4,8,1/flixel/group/FlxGroup.hx:473: characters 80-91 : Type not found : ObjectClass
/home/cloavii/haxelib/polymod/git/polymod/format/XMLMerge.hx:189: characters 10-23 : Warning : This typedef is deprecated in favor of haxe.xml.Access
/home/cloavii/haxelib/polymod/git/polymod/format/XMLMerge.hx:189: characters 30-43 : Warning : This typedef is deprecated in favor of haxe.xml.Access
/home/cloavii/haxelib/flixel/4,8,1/flixel/group/FlxGroup.hx:852: characters 19-30 : Type not found : ObjectClass
/home/cloavii/haxelib/flixel/4,8,1/flixel/system/frontEnds/PluginFrontEnd.hx:47: characters 19-28 : Type not found : ClassType
/home/cloavii/haxelib/flixel/4,8,1/flixel/system/frontEnds/PluginFrontEnd.hx:94: characters 20-29 : Type not found : ClassType

**Comment by DjBRINE** [comment]
What version of Haxe do you use? Try 4.1.5 instead of 4.2.x

**Comment by cloavii** [comment]
im unsure, is there a way to check that?

**Comment by cloavii** [comment]
nvm, turns out i have 4.2.1, how do i get 4.1.5?

**Comment by exurd** [comment]
Haxe 4.15 is available [here](https://haxe.org/download/version/4.1.5/).

Please follow the [build instructions](https://github.com/ninjamuffin99/Funkin#build-instructions) in the readme for more info. It will guide you on how to install required libraries and programs.

**Comment by cloavii** [comment]
when i tried to do that, it just downloaded the latest version

**Comment by exurd** [comment]
Did you uninstall the latest version before installing 4.1.5?

**Comment by cloavii** [comment]
i did, at least i think? i used the "sudo apt remove haxe" command in the terminal

**Comment by exurd** [comment]
Try `sudo apt install haxe=1:4.0.5-1`

**Comment by cloavii** [comment]
now im getting a different error

Warning: Lime version 7.9.0 may not be compatible with OpenFL 9.0.2 (expected version 7.7.*)
Use <lime version="" /> before <haxelib name="openfl" /> to include a specific version
source/Controls.hx:411: characters 25-33 : Array<Int> has no field contains
/home/cloavii/haxelib/polymod/git/polymod/format/XMLMerge.hx:189: characters 10-23 : Warning : This typedef is deprecated in favor of haxe.xml.Access
/home/cloavii/haxelib/polymod/git/polymod/format/XMLMerge.hx:189: characters 30-43 : Warning : This typedef is deprecated in favor of haxe.xml.Access

**Comment by exurd** [comment]
Seem to be having the same issue as #584. Try running `haxelib install lime 7.8.0`

**Comment by cloavii** [comment]
that fixed the issue, but now its just this part of the error message 

source/Controls.hx:411: characters 25-33 : Array<Int> has no field contains
/home/cloavii/haxelib/polymod/git/polymod/format/XMLMerge.hx:189: characters 10-23 : Warning : This typedef is deprecated in favor of haxe.xml.Access
/home/cloavii/haxelib/polymod/git/polymod/format/XMLMerge.hx:189: characters 30-43 : Warning : This typedef is deprecated in favor of haxe.xml.Access

**Comment by exurd** [comment]
Try going through the [build instructions](https://github.com/ninjamuffin99/Funkin#build-instructions) again. Maybe after uninstalling and reinstalling haxe, the packages have been removed as well.

**Comment by cloavii** [comment]
it looks like polymod is the problem here. i reinstalled it, and it still doesnt work

**Comment by cloavii** [comment]
update: just checked, and haxe 4.0.5 installed instead of 4.1.5

**Comment by itskwartz** [comment]
What command are you using to compile it?

**Comment by cloavii** [comment]
i tired both "lime test linux -debug" and "lime test linux -release", both got the same results

**Comment by exurd** [comment]
> update: just checked, and haxe 4.0.5 installed instead of 4.1.5

for some reason, haxe 4.1.5 isn't available on the repository:

```
$ apt-cache madison haxe
      haxe | 1:4.2.1-1~ubuntu20.04.1~ppa1 | http://ppa.launchpad.net/haxe/releases/ubuntu focal/main amd64 Packages
      haxe |  1:4.0.5-1 | http://archive.ubuntu.com/ubuntu focal/universe amd64 Packages
```

**Comment by cloavii** [comment]
damn, i think ill save myself the trouble and just use wine instead, thank you though!

**Comment by cloavii** [comment]
update: nevermind, even wine has some issues.

**Comment by cloavii** [comment]
is there a way to get haxe 4.1.5, even with the repo not having it?

**Comment by exurd** [comment]
You can get 4.1.5 by downloading the Linux binaries on the download page, but I don't know what to do after that.

**Comment by cloavii** [comment]
i dont know what to do with these files either, lets hope someone who knows this stuff can come along and help

**Comment by exurd** [comment]
I have searched on how to use the binaries and [found this](https://stackoverflow.com/questions/58052420/haxe-installation-from-binaries). Try it if you want to.

**Comment by DjBRINE** [comment]
Debian repository have haxe 4.1.5 availiable. Here's the direct link from debian repo
http://ftp.br.debian.org/debian/pool/main/h/haxe/haxe_4.1.5-1_amd64.deb
And a site where i got this info, so you can verify
https://debian.pkgs.org/sid/debian-main-amd64/haxe_4.1.5-1_amd64.deb.html

**Comment by DjBRINE** [comment]
BTW, you will experience some problems even if you compile it from source code, like updated out-of-date version, etc.
If freeplay decides to crash the game, you can find solution here https://github.com/ninjamuffin99/Funkin/issues/501#issuecomment-787261843

**Comment by AlekEagle** [comment]
I've done everything so far, unfortunately, running trying to build for any platform causes lime to eventually close with the only output being `Type not found : haxe.Exception` did I mess up my install?

**Comment by AlekEagle** [comment]
Ok, I've gotten a bit further, I did mess up my install, I merged `/usr/share/haxe/std` with the one that was left by my initial install, and that caused some files to not get copied into the folder, I completely deleted `/usr/share/haxe/std` and copied the 4.1.5 std folder into it, which fixed it. The only issue I'm getting now is neko getting stuck on building `src/flixel/input/actions/FlxActionInputAnalogSteam.cpp` and causing a memory leak. Any help would be greatly appreciated.

**Comment by AlekEagle** [comment]
I was able to successfully compile the game with discord_rpc! I had to copy the hxcpp.h file from the hxcpp lib in the haxelib directory to the appropriate __pch haxe directory. here's how I was able to successfully build everything.

1. Begin compilation with your `lime build linux` you can include the `-debug` flag if you want, and allow it to error at \_\_boot\_\_.cpp with it saying something like 
```
Error: In file included from include/discord_rpc/DiscordRpc.h:11:0,
from ./src/__boot__.cpp:3449:
/home/alekeagle/haxelib/discord_rpc/git/linc/linc_discord_rpc.h:6:10: fatal error: /drives/4tb/projects/Funkin/export/release/linux/obj/obj/linux64/__pch/haxe/hxcpp.h:No such file or directory
#include <hxcpp.h>
^~~~~~~~~
compilation terminated.
```
2. Locate your haxelib folder, and copy hxcpp.h to the directory that the aforementioned error said the file couldn't be located at.
3. Rerun compilation and your build should complete successfully. It looks like this is an error with discord_rpc, so the fix should be made in discord_rpc.

**Comment by cloavii** [comment]
sorry for not updating the thread, but thank you! ill try this when i get home

**Comment by cloavii** [comment]
> I was able to successfully compile the game with discord_rpc! I had to copy the hxcpp.h file from the hxcpp lib in the haxelib directory to the appropriate __pch haxe directory. here's how I was able to successfully build everything.
> 
> 1. Begin compilation with your `lime build linux` you can include the `-debug` flag if you want, and allow it to error at __boot__.cpp with it saying something like
> 
> ```
> Error: In file included from include/discord_rpc/DiscordRpc.h:11:0,
> from ./src/__boot__.cpp:3449:
> /home/alekeagle/haxelib/discord_rpc/git/linc/linc_discord_rpc.h:6:10: fatal error: /drives/4tb/projects/Funkin/export/release/linux/obj/obj/linux64/__pch/haxe/hxcpp.h:No such file or directory
> #include <hxcpp.h>
> ^~~~~~~~~
> compilation terminated.
> ```
> 
> 1. Locate your haxelib folder, and copy hxcpp.h to the directory that the aforementioned error said the file couldn't be located at.
>    Here are some scripts you can use to quickly copy the header file to the proper location (run these in the root folder of Funkin):
> 
> ```shell
> # when compiling release build
> cp $(haxelib config)hxcpp/$(sed -e 's:\.:,:g' $(haxelib config)hxcpp/.current)/include/hxcpp.h export/release/linux/obj/obj/linux64/__pch/haxe/hxcpp.h
> 
> # when compiling debug build
> cp $(haxelib config)hxcpp/$(sed -e 's:\.:,:g' $(haxelib config)hxcpp/.current)/include/hxcpp.h export/debug/linux/obj/obj/linux64-debug/__pch/haxe/hxcpp.h
> ```
> 
> 1. Rerun compilation and your build should complete successfully. It looks like this is an error with discord_rpc, so the fix should be made in discord_rpc.

thank you so much for this! it literally fixed the whole thing :)

**Comment by OPNA2608** [comment]
Opened an issue on discord_rpc about it. FWIW the header problem is fixable by re-adding the following to discord_rpc's `linc/linc_discord_rpc-xml`.
```xml
   <section id = "vars">

     <set name = "NO_PRECOMPILED_HEADERS" value = "1" />

   </section>
```

In case anyone else encounters this: I had another problem with discord_rpc afterwards (Haxe 4.2.1 if it matters) where the `onRequest` method failed to compile like so:

> ```
> Error: ./src/discord_rpc/_DiscordRpc/DiscordRpc_Impl_.cpp: In static member function 'static void discord_rpc::_DiscordRpc::DiscordRpc_Impl__obj::_onRequest(const DiscordJoinRequest*)':
> ./src/discord_rpc/_DiscordRpc/DiscordRpc_Impl_.cpp:50:64: error: no match for call to '(Dynamic) (DiscordJoinRequest&)'
>    50 | HXDLIN( 161)   ::discord_rpc::DiscordRpc_obj::onRequest(_hx_tmp);
>       |                                                                ^
> In file included from /nix/store/v8ynxhz2n58h96rwya65pi5mr6c2rd8n-hxcpp-4.2.1/lib/haxe/hxcpp/4,2,1/include/hxcpp.h:353:
> /nix/store/v8ynxhz2n58h96rwya65pi5mr6c2rd8n-hxcpp-4.2.1/lib/haxe/hxcpp/4,2,1/include/Dynamic.h:304:22: note: candidate: 'Dynamic Dynamic::operator()()'
>   304 |    inline  ::Dynamic operator()() { CheckFPtr(); return mPtr->__run(); }
>       |                      ^~~~~~~~
> /nix/store/v8ynxhz2n58h96rwya65pi5mr6c2rd8n-hxcpp-4.2.1/lib/haxe/hxcpp/4,2,1/include/Dynamic.h:304:22: note:   candidate expects 0 arguments, 1 provided
> /nix/store/v8ynxhz2n58h96rwya65pi5mr6c2rd8n-hxcpp-4.2.1/lib/haxe/hxcpp/4,2,1/include/Dynamic.h:305:22: note: candidate: 'Dynamic Dynamic::operator()(const Dynamic&)'
>   305 |    inline  ::Dynamic operator()(const Dynamic &inArg0) { CheckFPtr(); return mPtr->__run(inArg0); }
>       |                      ^~~~~~~~
> /nix/store/v8ynxhz2n58h96rwya65pi5mr6c2rd8n-hxcpp-4.2.1/lib/haxe/hxcpp/4,2,1/include/Dynamic.h:305:48: note:   no known conversion for argument 1 from 'DiscordJoinRequest' to 'const Dynamic&'
>   305 |    inline  ::Dynamic operator()(const Dynamic &inArg0) { CheckFPtr(); return mPtr->__run(inArg0); }
>       |
> ```

I fixed this on my end by searching through `discord_rpc/DiscordRpc.hx`, replacing `JoinRequest->Void` with `ConstPointer<JoinRequest>->Void` and `ptr.value` with `ptr`. I assume this compiles & runs fine because `onRequest` isn't actually used by this project, but it's prolly not a proper fix for the problem? ¯\\_(ツ)_/¯

**Comment by OPNA2608** [comment]
Should be fixed properly now, I don't encounter either of those problems anymore.

**Comment by Jaqueka1432** [comment]
i read something about installing haxe 4.1.5 from binaries???
well I managed to do it easy and simple
1- go to haxe.org/download/version/4.1.5 and download the binaries (tar.gz files)
2- extract on desktop or somewhere easily accessible
3- copy the haxe, haxelib files and std folder to /bin
4- congratulations you just installed haxe 4.1.5

**Comment by VMGuy23** [comment]
I compiled fine with Haxe 4.2.1

**Comment by aflacc** [comment]
longest issue

**Comment by VMGuy23** [comment]
Yeah
Whatever is causing issue, I did the optional fix for zoomed out, using Haxe 4.2.1
